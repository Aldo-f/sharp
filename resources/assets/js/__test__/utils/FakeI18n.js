export default {
    install() {
        window.i18n = {};
    }
}