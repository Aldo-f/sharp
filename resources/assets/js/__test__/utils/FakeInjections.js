import Vue from 'vue';

const injectedComponents = ['$field', '$form'];

export default {
    provide() {
        return injectedComponents.reduce((injections, prop) => {
            injections[prop] = new Vue();
            return injections;
        }, {});
    }
}