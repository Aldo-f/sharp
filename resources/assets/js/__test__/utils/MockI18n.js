import Localization ,* as localizationFn from '../../mixins/Localization';

const mockLangImplementation = localeKey => `{{ ${localeKey} }}`;

export default {
    install() {
        window.i18n = {};
    },

    mockLangFunction() {
        localizationFn.lang = jest.fn(mockLangImplementation);
        Localization.methods.l = jest.fn(mockLangImplementation);
    }
}