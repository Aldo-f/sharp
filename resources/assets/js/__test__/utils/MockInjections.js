import Vue from 'vue';

const injectedComponents = {
    $field:{},
    $form:{
        data: () => ({
            errors: {}
        })
    },
    $tab:{}
};

export default {
    provide() {
        return Object.keys(injectedComponents).reduce((injections, compName) => {
            let Comp = Vue.extend(injectedComponents[compName]);
            injections[compName] = new Comp();
            return injections;
        }, {});
    }
}