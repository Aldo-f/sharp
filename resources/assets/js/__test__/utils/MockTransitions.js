export default {
    install() {
        window.getComputedStyle = jest.fn(() => ({
            animationDelay:'',
            animationDuration:'',
            transitionDelay:'',
            transitionDuration:''
        }));
    }
}