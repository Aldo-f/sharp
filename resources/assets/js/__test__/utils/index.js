export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC } from './mockSFC';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}