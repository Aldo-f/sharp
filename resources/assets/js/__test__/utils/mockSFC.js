export function mockSFC (sfcModule, { template, render, ...extend }={}, mockOptions={}) {

    let { isJsFile } = mockOptions;

    if(isJsFile) {
        sfcModule.__oldDefault__ = { ...sfcModule.default }
    }

    Object.assign(sfcModule.default, {
        mixins: [],
        'extends': extend,
        template: template,
        render: !template ? render || function() { return this._v(`__MOCKED_SFC_${sfcModule.name}__\n`) } : null,
        watch: {},
        beforeCreate: ()=>{},
        created: ()=>{},
        beforeMount: ()=>{},
        mounted: ()=>{},
        beforeUpdate: ()=>{},
        updated: ()=>{},
        beforeDestroy: ()=>{},
        destroyed: ()=>{},
        data: ()=>({}),
        provide: {},
        inject: [],
    });

    //console.log(sfcModule);
    return sfcModule.default;
}

export function unmockSFC(sfcModule) {
    if(sfcModule.__oldDefault__) {
        sfcModule.default = sfcModule.__oldDefault__;
        return;
    }

    let { 'default':defaultObj, ...options } = sfcModule;
    sfcModule.default = options;
}