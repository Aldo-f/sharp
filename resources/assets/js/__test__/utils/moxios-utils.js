import moxios from 'moxios';

export function nextRequestFulfilled(config) {
    return new Promise(resolve => {
        moxios.wait(async () => {
            let request = moxios.requests.mostRecent();
            let response = await request.respondWith(config);
            resolve(response); // moxios Response object
        })
    });
}
