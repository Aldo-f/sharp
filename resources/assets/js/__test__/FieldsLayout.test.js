import Vue from 'vue';
import FieldsLayout from '../components/form/FieldsLayout.vue';

import * as gridModule from '../components/Grid.vue';

import { MockInjections, mockSFC, unmockSFC } from "./utils";

describe('fields-layout', () => {
    Vue.component('sharp-fields-layout', FieldsLayout);

    beforeAll(()=>{
        mockSFC(gridModule,{
            template: `
            <div id="MOCKED_GRID">
                <slot v-bind="rows[0][0]"></slot>
            </div>
            `
        });
    });

    afterAll(()=>{
        unmockSFC(gridModule);
    });

    beforeEach(()=>{
        Vue.component('field-mock', {
            render() { return this._v('FIELD MOCK'); }
        });

        document.body.innerHTML = `
            <div id="app">
                <sharp-fields-layout :layout="layout" :visible="visible">
                    <template slot-scope="fieldLayout">
                        <field-mock :data="fieldLayout" ref="fieldMock"></field-mock>           
                    </template>
                </sharp-fields-layout>      
            </div>
        `
    });

    it('can mount fields layout', async () => {
        await createVm({
            propsData: {
                layout: [
                    [{ key:'title' }]
                ]
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can mount "fieldset" fields layout', async () => {
        await createVm({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset', fields: [[
                            { key: 'title' }
                        ]]
                    }]
                ],
                visible: {
                    title: true
                }
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can mount "hidden fieldset" fields layout', async () => {
        await createVm({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset', fields: [
                            [{ key: 'title' }]
                        ]
                    }]
                ],
                visible: {
                    title: false
                }
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('expose correct props', async () => {
        let { $root:vm } = await createVm({
            propsData: {
                layout: [
                    [{ key:'list' }]
                ]
            }
        });

        expect(vm.$refs.fieldMock.$attrs).toEqual({
            data: { key: 'list' }
        });
    });

    it('expose correct fieldset props', async () => {
        let { $root:vm } = await createVm({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset', fields: [
                            [{ key: 'title' }]
                        ]
                    }]
                ],
            }
        });

        expect(vm.$refs.fieldMock.$attrs).toEqual({
            data: { key: 'title' }
        });
    });

    it('fieldset visible', async () => {
        let $fieldsLayout = await createVm({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset 1',
                        id: 'fieldset_1',
                        fields: [
                            [{ key: 'title' }, { key: 'subtitle' }],[{ key: 'name' }]
                        ]
                    }]
                ],
                visible: {
                    title: true, subtitle: true, name: false,
                }
            }
        });

        expect($fieldsLayout.fieldsetMap).toEqual({
            'fieldset_1':[{ key: 'title' }, { key: 'subtitle' },{ key: 'name' }]
        });

        expect($fieldsLayout.isFieldsetVisible({ id: 'fieldset_1'})).toBe(true);
    });

    it('fieldset invisible', async () => {
        let $fieldsLayout = await createVm({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset 1',
                        id: 'fieldset_1',
                        fields: [
                            [{ key: 'title' }, { key: 'subtitle' }],[{ key: 'name' }]
                        ]
                    }]
                ],
                visible: {
                    title: false, subtitle: false, name: false,
                }
            }
        });

        expect($fieldsLayout.fieldsetMap).toEqual({
            'fieldset_1':[{ key: 'title' }, { key: 'subtitle' },{ key: 'name' }]
        });

        expect($fieldsLayout.isFieldsetVisible({ id: 'fieldset_1'})).toBe(false);
    })

});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [MockInjections, customOptions],

        props: ['layout', 'visible']
    });

    await Vue.nextTick();

    return vm.$children[0];
}