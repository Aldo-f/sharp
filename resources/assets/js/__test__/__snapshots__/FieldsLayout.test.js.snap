// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fields-layout can mount "fieldset" fields layout 1`] = `
<div id="app">
	<div id="MOCKED_GRID">
		<fieldset class="SharpForm__fieldset">
			<div class="SharpModule__inner">
				<div class="SharpModule__header">
					<legend class="SharpModule__title">Fieldset</legend>
				</div>
				<div class="SharpModule__content">
					<div id="MOCKED_GRID">FIELD MOCK</div>
				</div>
			</div>
		</fieldset>
	</div>
</div>
`;

exports[`fields-layout can mount "hidden fieldset" fields layout 1`] = `
<div id="app">
	<div id="MOCKED_GRID">
		<fieldset class="SharpForm__fieldset" style="display: none;">
			<div class="SharpModule__inner">
				<div class="SharpModule__header">
					<legend class="SharpModule__title">Fieldset</legend>
				</div>
				<div class="SharpModule__content">
					<div id="MOCKED_GRID">FIELD MOCK</div>
				</div>
			</div>
		</fieldset>
	</div>
</div>
`;

exports[`fields-layout can mount fields layout 1`] = `
<div id="app">
	<div id="MOCKED_GRID">FIELD MOCK</div>
</div>
`;
